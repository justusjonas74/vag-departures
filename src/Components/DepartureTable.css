.deapartureContainer h3 {
    line-height: 1;
    margin-bottom: 1rem;
}