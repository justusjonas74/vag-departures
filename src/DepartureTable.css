.departureRow .departureLine {
    margin-left: 3px;
    color: white;
    font-weight: bold;
    padding: 1px 5px;
    border-radius: 5px;
}

.departureRow .departureLine.ubahn {
    background-color: #006291;
}

.departureRow .departureLine.tram {
    background-color: #852D6D;
}

.departureRow .departureLine.bus {
    background-color: #E4342E;
}

.departureRow .departureTime {
    font-size: 1.2rem;
    font-weight: bold;
}

.departureRow .departureDelay {
    padding-left: 0.3rem;
    font-size: 0.8rem;
}



.departureRow .departureDelay.isDelayed {
    color: red;
    font-weight: bold;
}

.departureRow .departureDelay.isInTime {
    color: green;
}

.deapartureContainer h3 {
    line-height: 1;
    margin-bottom: 1rem;
}