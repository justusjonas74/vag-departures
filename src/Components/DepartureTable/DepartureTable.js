import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './DepartureTable.css';
import DepartureRow from '../DepartureRow/DepartureRow';
// var strftime = require('strftime');
import moment from 'moment'
// TODO: REPLACE strftime with moment.js!

/* ----------------- */
/*    Helper Methods */
/* ----------------- */
function convertHourMinute(value) {
  // const date = new Date(value);
  // const date = moment(value);
  // if (date) {
  //   return strftime('%H:%M', date);
  // }
  // else {
  //   return '?'
  // }
  return moment(value).format('HH:mm')
}

function checkDepartureDelay(datePlannedStr, dateRealStr) {
  const datePlanned = new Date(datePlannedStr),
    dateReal = new Date(dateRealStr);
  const delay = dateReal - datePlanned;
  var ONE_MINUTE = 60 * 1000; /* ms */
  if (delay < ONE_MINUTE) {
    return {
      isDelayed: false,
      time: 0
    }
  }
  else {
    const output = Math.floor(delay / ONE_MINUTE);
    return {
      isDelayed: true,
      time: output
    }
  }
}

/* ----------------- */
/*    EXPORTS        */
/* ----------------- */

class DepartureTable extends Component {
  // <i className="fa fa-map-marker" aria-hidden="true"></i>
  render() {
    const rows = [];
    const departures = this.props.departures ? this.props.departures : [];
    const stopName = this.props.Haltestellenname ? this.props.Haltestellenname : ""
    if (typeof departures !== 'undefined' && departures.length > 0) {
      departures.forEach((item, index) => {
        const delay = checkDepartureDelay(item.AbfahrtszeitSoll, item.AbfahrtszeitIst);
        rows.push(
          <tr className="departureRow" key={index}>
            <DepartureRow
              departureTime={convertHourMinute(item.AbfahrtszeitSoll)}
              delay={delay}
              departureDirection={item.Richtungstext}
              departureLine={item.Linienname}
              departureTransitType={item.Produkt}
              key={index}/>
          </tr>
        )
      })
      return (
        <div className="deapartureContainer">
            <div className='btn-toolbar pull-right'>
              <div className='btn-group'>
                <button type='button' onClick={this.props.newSearchButtonFn} className='btn btn-outline-secondary btn-sm'><i className="fa fa-search"></i> New Search</button>
              </div>
            </div>
            <h3>{stopName}</h3>
            <table className="table table-striped<">
              <thead>
                <tr>
                  <th>Time</th>
                  <th>Line</th>
                  <th>Direction</th>
                </tr>
              </thead>
              <tbody>
                {rows}
              </tbody>
            </table>
          </div>
      )
    } else {
      return null
    }
  }
}

DepartureTable.propTypes = {
  departures: PropTypes.array,
  Haltestellenname: PropTypes.string,
  newSearchButtonFn: PropTypes.func 
}

export default DepartureTable