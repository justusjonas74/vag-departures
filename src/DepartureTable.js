import React, { Component } from 'react';
import './DepartureTable.css';
var strftime = require('strftime');
// TODO: REPLACE strftime with moment.js!

/* ----------------- */
/*    Helper Methods */
/* ----------------- */
function convertHourMinute(value) {
    const date = new Date(value);
    if (date) {
        return strftime('%H:%M', date);
    }
    else {
        return '?'
    }
}


function checkDepartureDelay(datePlannedStr, dateRealStr) {
    const datePlanned = new Date(datePlannedStr),
        dateReal = new Date(dateRealStr);
    const delay = dateReal - datePlanned;
    var ONE_MINUTE = 60 * 1000; /* ms */
    if (delay < ONE_MINUTE) {
        return {
            isDelayed: false,
            time: 0
        }
    }
    else {
        const output = Math.floor(delay / ONE_MINUTE);
        return {
            isDelayed: true,
            time: output
        }
    }

}


/* ----------------- */
/*    SUB COMPONENTS */
/* ----------------- */
class LineTransitTypeComponent extends Component {
  styleTransitType(departureTransitType) {
    switch (departureTransitType.toLowerCase()) {
      case 'ubahn':
        return {
          cssClass: 'ubahn',
          symbolStr: 'fa-train'
        };
      case 'tram':
        return {
          cssClass: 'tram',
          symbolStr: 'fa-subway'
        };
      case 'bus':
        return {
          cssClass: 'bus',
          symbolStr: 'fa-bus'
        };
      default:
        return {
          cssClass: 'bus',
          symbolStr: 'fa-bus'
        };
    }
  }
    
    render() {
        const departureLine = this.props.departureLine
        const departureTransitType = this.props.departureTransitType;
        const typeStyles = this.styleTransitType(departureTransitType);

        return (
            <td>
        <span className="departureTransitType"><i className={'fa ' + typeStyles.symbolStr} aria-hidden="true"></i>
        <span className={'departureLine '+ typeStyles.cssClass}>{departureLine}</span></span>
      </td>
        )
    }
}


class DepartureRow extends Component {

    render() {
        return (
            <tr className="departureRow">
          <td>
            <span className="departureTime">{this.props.departureTime}</span>
            <span className={"departureDelay " + (this.props.delay.isDelayed ? 'isDelayed' : 'isInTime')} >+{this.props.delay.time} </span></td>
          <LineTransitTypeComponent
            departureLine={this.props.departureLine}
            departureTransitType={this.props.departureTransitType}
          />
          <td><span className="departureDirection">{this.props.departureDirection}</span></td>
        </tr>
        )
    }
}


/* ----------------- */
/*    EXPORTS        */
/* ----------------- */



class DepartureTable extends Component {

    render() {
        const rows = [];
        const departures = this.props.departures;

        departures.forEach((item, index) => {
            const delay = checkDepartureDelay(item.AbfahrtszeitSoll, item.AbfahrtszeitIst);
            rows.push(
                <DepartureRow
            departureTime={convertHourMinute(item.AbfahrtszeitSoll)}
            delay={delay}
            departureDirection={item.Richtungstext}
            departureLine={item.Linienname}
            departureTransitType={item.Produkt}
            key={index}/>
            )
        })
        if (departures.length < 1) {
            return null
        }
        else {
            return (
                <div className="deapartureContainer">
            <table className="table table-striped<">
              <thead>
                <th>Time</th>
                <th>Line</th>
                <th>Direction</th>
              </thead>
              <tbody>
                {rows}
              </tbody>
            </table>
          </div>
            )
        }

    }
}

export default DepartureTable