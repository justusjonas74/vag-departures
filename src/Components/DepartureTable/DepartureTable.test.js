import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import DepartureTable from './DepartureTable';


const response = {
  "Metadata": {
    "Version": "Puls-API-v1.1",
    "Timestamp": "2018-11-16T13:07:09+01:00"
  },
  "Haltestellenname": "Bauernfeindstr. (Nürnberg)",
  "VAGKennung": "BAUERN,BA",
  "VGNKennung": 1550,
  "Abfahrten": [{
    "Linienname": "U1",
    "Haltepunkt": "BA:2",
    "Richtung": "Richtung1",
    "Richtungstext": "Langwasser Süd",
    "AbfahrtszeitSoll": "2018-11-16T13:11:29+01:00",
    "AbfahrtszeitIst": "2018-11-16T13:11:29+01:00",
    "Produkt": "UBahn",
    "Longitude": 11.10790611,
    "Latitude": 49.41769389,
    "Fahrtnummer": 1005407,
    "Fahrtartnummer": 1,
    "Prognose": false
  }, {
    "Linienname": "U1",
    "Haltepunkt": "BA:1",
    "Richtung": "Richtung2",
    "Richtungstext": "Fürth Hardhöhe",
    "AbfahrtszeitSoll": "2018-11-16T13:12:07+01:00",
    "AbfahrtszeitIst": "2018-11-16T13:12:07+01:00",
    "Produkt": "UBahn",
    "Longitude": 11.10790611,
    "Latitude": 49.41769389,
    "Fahrtnummer": 1005744,
    "Fahrtartnummer": 1,
    "Prognose": false
  }, {
    "Linienname": "93",
    "Haltepunkt": "BAUERN:6",
    "Richtung": "Richtung1",
    "Richtungstext": "Kornburg",
    "AbfahrtszeitSoll": "2018-11-16T13:16:00+01:00",
    "AbfahrtszeitIst": "2018-11-16T13:16:00+01:00",
    "Produkt": "Bus",
    "Longitude": 11.10790611,
    "Latitude": 49.41769389,
    "Fahrtnummer": 2039443,
    "Fahrtartnummer": 1,
    "Fahrzeugnummer": "551",
    "Prognose": true
  }]
}
describe('DepartureTable Component', () => {
  
  // BOILERPLATE
  let props;
  let mountedComponent;
  const departureTable = () => {
    if (!mountedComponent) {
      // mountedComponent = mount(
      mountedComponent = mount(
        <DepartureTable {...props} />
      );
    }
    return mountedComponent;
  }

  beforeEach(() => {
    props = {
      departures: null,
      Haltestellenname: null,
      newSearchButtonFn: null 
    };
    mountedComponent = undefined;
  });
  
  it('should return nothing, if props.departures is null', ()=>{
    const component = departureTable().html()
    
    // const component = shallow(<DepartureTable {...props} />)
    // expect(component.get(0)).toBeFalsy()
    expect(component).toBeNull();
  })
  
  describe('with props', () => {
    beforeEach(()=>{
      props.departures = response.Abfahrten
      props.Haltestellenname = response.Haltestellenname
    })
    
    it('should render correctly', () => {
       const tree = renderer
        .create(<DepartureTable {...props} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    
  })
  
})




// TEST DATA

