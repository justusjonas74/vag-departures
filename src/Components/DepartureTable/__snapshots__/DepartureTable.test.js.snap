// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepartureTable Component with delay should render correctly 1`] = `
<div
  className="deapartureContainer"
>
  <div
    className="btn-toolbar pull-right"
  >
    <div
      className="btn-group"
    >
      <button
        className="btn btn-outline-secondary btn-sm"
        onClick={null}
        type="button"
      >
        <i
          className="fa fa-search"
        />
         New Search
      </button>
    </div>
  </div>
  <h3>
    Bauernfeindstr. (Nürnberg)
  </h3>
  <table
    className="table table-striped<"
  >
    <thead>
      <tr>
        <th>
          Time
        </th>
        <th>
          Line
        </th>
        <th>
          Direction
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="departureRow"
      >
        <td>
          <span
            className="departureTime"
          >
            13:11
          </span>
          <span
            className="departureDelay isDelayed"
          >
            +
            4
          </span>
        </td>
        <td>
          <span
            className="departureTransitType"
          >
            <i
              aria-hidden="true"
              className="fa fa-train"
            />
            <span
              className="departureLine ubahn"
            >
              U1
            </span>
          </span>
        </td>
        <td>
          <span
            className="departureDirection"
          >
            Langwasser Süd
          </span>
        </td>
      </tr>
      <tr
        className="departureRow"
      >
        <td>
          <span
            className="departureTime"
          >
            13:12
          </span>
          <span
            className="departureDelay isDelayed"
          >
            +
            20
          </span>
        </td>
        <td>
          <span
            className="departureTransitType"
          >
            <i
              aria-hidden="true"
              className="fa fa-train"
            />
            <span
              className="departureLine ubahn"
            >
              U1
            </span>
          </span>
        </td>
        <td>
          <span
            className="departureDirection"
          >
            Fürth Hardhöhe
          </span>
        </td>
      </tr>
      <tr
        className="departureRow"
      >
        <td>
          <span
            className="departureTime"
          >
            13:16
          </span>
          <span
            className="departureDelay isDelayed"
          >
            +
            60
          </span>
        </td>
        <td>
          <span
            className="departureTransitType"
          >
            <i
              aria-hidden="true"
              className="fa fa-bus"
            />
            <span
              className="departureLine bus"
            >
              93
            </span>
          </span>
        </td>
        <td>
          <span
            className="departureDirection"
          >
            Kornburg
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`DepartureTable Component with props (without delayed trips) should render correctly 1`] = `
<div
  className="deapartureContainer"
>
  <div
    className="btn-toolbar pull-right"
  >
    <div
      className="btn-group"
    >
      <button
        className="btn btn-outline-secondary btn-sm"
        onClick={null}
        type="button"
      >
        <i
          className="fa fa-search"
        />
         New Search
      </button>
    </div>
  </div>
  <h3>
    Bauernfeindstr. (Nürnberg)
  </h3>
  <table
    className="table table-striped<"
  >
    <thead>
      <tr>
        <th>
          Time
        </th>
        <th>
          Line
        </th>
        <th>
          Direction
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="departureRow"
      >
        <td>
          <span
            className="departureTime"
          >
            13:11
          </span>
          <span
            className="departureDelay isInTime"
          >
            +
            0
          </span>
        </td>
        <td>
          <span
            className="departureTransitType"
          >
            <i
              aria-hidden="true"
              className="fa fa-train"
            />
            <span
              className="departureLine ubahn"
            >
              U1
            </span>
          </span>
        </td>
        <td>
          <span
            className="departureDirection"
          >
            Langwasser Süd
          </span>
        </td>
      </tr>
      <tr
        className="departureRow"
      >
        <td>
          <span
            className="departureTime"
          >
            13:12
          </span>
          <span
            className="departureDelay isInTime"
          >
            +
            0
          </span>
        </td>
        <td>
          <span
            className="departureTransitType"
          >
            <i
              aria-hidden="true"
              className="fa fa-train"
            />
            <span
              className="departureLine ubahn"
            >
              U1
            </span>
          </span>
        </td>
        <td>
          <span
            className="departureDirection"
          >
            Fürth Hardhöhe
          </span>
        </td>
      </tr>
      <tr
        className="departureRow"
      >
        <td>
          <span
            className="departureTime"
          >
            13:16
          </span>
          <span
            className="departureDelay isInTime"
          >
            +
            0
          </span>
        </td>
        <td>
          <span
            className="departureTransitType"
          >
            <i
              aria-hidden="true"
              className="fa fa-bus"
            />
            <span
              className="departureLine bus"
            >
              93
            </span>
          </span>
        </td>
        <td>
          <span
            className="departureDirection"
          >
            Kornburg
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
